src/easyscience/fitting/minimizers/minimizer_dfo.py

```diff
@@ -211,19 +116,115 @@ def fit(
+        :rtype: Callable
+        """
+        fit_func = self._generate_fit_function()
+
+        def _outer(obj: DFO):
+            def _make_func(x, y, weights):
+                ## TODO clean when full move to new_variable
+                from easyscience.Objects.new_variable import Parameter as NewParameter
+
+                pars = {}
+                if not parameters:
+                    for name, par in obj._cached_pars.items():
+                        if isinstance(par, NewParameter):
+                            pars[MINIMIZER_PARAMETER_PREFIX + str(name)] = par.value
+                        else:
+                            pars[MINIMIZER_PARAMETER_PREFIX + str(name)] = par.raw_value
+
+                else:
+                    for new_par in parameters:
+                        if isinstance(new_par, NewParameter):
+                            pars[MINIMIZER_PARAMETER_PREFIX + new_par.unique_name] = new_par.value
+                        else:
+                            pars[MINIMIZER_PARAMETER_PREFIX + new_par.unique_name] = new_par.raw_value
+
+                def _residuals(pars_values: List[float]) -> np.ndarray:
+                    for idx, par_name in enumerate(pars.keys()):
+                        pars[par_name] = pars_values[idx]
+                    return (y - fit_func(x, **pars)) / weights
+
+                return _residuals
+
+            return _make_func
+
+        return _outer(self)
+
+    def _generate_fit_function(self) -> Callable:
+        """
+        Using the user supplied `fit_function`, wrap it in such a way we can update `Parameter` on
+        iterations.
+
+        :return: a fit function which is compatible with bumps models
+        :rtype: Callable
+        """
+        # Original fit function
+        func = self._original_fit_function
+        # Get a list of `Parameters`
+        self._cached_pars = {}
+        self._cached_pars_vals = {}
+        for parameter in self._object.get_fit_parameters():
+            key = parameter.unique_name
+            self._cached_pars[key] = parameter
+            self._cached_pars_vals[key] = (parameter.value, parameter.error)
+
+        # Make a new fit function
+        def _fit_function(x: np.ndarray, **kwargs):
+            """
+            Wrapped fit function which now has an EasyScience compatible form
+
+            :param x: array of data points to be calculated
+            :type x: np.ndarray
+            :param kwargs: key word arguments
+            :return: points calculated at `x`
+            :rtype: np.ndarray
+            """
+            # Update the `Parameter` values and the callback if needed
+            # TODO THIS IS NOT THREAD SAFE :-(
+            for name, value in kwargs.items():
+                par_name = name[1:]
+                if par_name in self._cached_pars.keys():
+                    ## TODO clean when full move to new_variable
+                    from easyscience.Objects.new_variable import Parameter
+
```

Comment:
you can move this import outside the loop

Reply:
changed